apis/v1beta1/config.go

```diff
@@ -445,64 +441,61 @@ const (
 // because the port is used to generate Service objects and mappings.
 func (s *Service) MetricsEndpoint(logger logr.Logger) (string, int32, error) {
 	telemetry := s.GetTelemetry()
-	if telemetry == nil || telemetry.Metrics.Address == "" {
+	if telemetry == nil {
 		return defaultServiceHost, defaultServicePort, nil
 	}
 
-	// The regex below matches on strings that end with a colon followed by the environment variable expansion syntax.
-	// So it should match on strings ending with: ":${env:POD_IP}" or ":${POD_IP}".
-	const portEnvVarRegex = `:\${[env:]?.*}$`
-	isPortEnvVar := regexp.MustCompile(portEnvVarRegex).MatchString(telemetry.Metrics.Address)
-	if isPortEnvVar {
-		errMsg := fmt.Sprintf("couldn't determine metrics port from configuration: %s",
-			telemetry.Metrics.Address)
-		logger.Info(errMsg)
-		return "", 0, errors.New(errMsg)
-	}
-
-	// The regex below matches on strings that end with a colon followed by 1 or more numbers (representing the port).
-	const explicitPortRegex = `:(\d+$)`
-	explicitPortMatches := regexp.MustCompile(explicitPortRegex).FindStringSubmatch(telemetry.Metrics.Address)
-	if len(explicitPortMatches) <= 1 {
-		return telemetry.Metrics.Address, defaultServicePort, nil
-	}
+	if telemetry.Metrics.Address != "" && len(telemetry.Metrics.Readers) == 0 {
+		host, port, err := parseAddressEndpoint(telemetry.Metrics.Address)
+		if err != nil {
+			return "", 0, err
+		}
 
-	port, err := strconv.ParseInt(explicitPortMatches[1], 10, 32)
-	if err != nil {
-		errMsg := fmt.Sprintf("couldn't determine metrics port from configuration: %s",
-			telemetry.Metrics.Address)
-		logger.Info(errMsg, "error", err)
-		return "", 0, err
+		return host, port, nil
+	}
+
+	port := defaultServicePort
+	host := defaultServiceHost
+	for _, reader := range telemetry.Metrics.Readers {
+		if reader.Pull != nil {
+			if reader.Pull.Exporter.Prometheus != nil {
+				if reader.Pull.Exporter.Prometheus.Host != nil && *reader.Pull.Exporter.Prometheus.Host != "" {
+					host = *reader.Pull.Exporter.Prometheus.Host
+				}
+				if reader.Pull.Exporter.Prometheus.Port != nil && *reader.Pull.Exporter.Prometheus.Port != 0 {
+					port = intToInt32Safe(int(*reader.Pull.Exporter.Prometheus.Port))
+				}
+				break
+			}
+		}
```

Comment:
In my opinion we should do this in a follow-up. It's a separate feature to generate Service port entries for all Pull readers. This implementation also only takes the first host/port pair, which is a bit arbitrary. I think we should drop this logic and open a separate issue to add it later.